### ISSUE TITLE
gmi sync hangs forever

### ISSUE THREAD

**Issue by kirelagin** [gauteh/lieer#150]
Possibly related to #83.

I am using it on a laptop with macOS. I have a launchd job that runs `gmi sync` quite frequently. Sometimes I find this job hanging, and it can remain in this state for _days_. I think this has something to do with it being a laptop that sometimes looses connectivity and/or goes to sleep often.

<details><summary>Backtrace</summary>

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff67df297e libsystem_kernel.dylib`read + 10
    frame #1: 0x000000010bfcc842 libcrypto.1.1.dylib`sock_read + 50
    frame #2: 0x000000010bfc7470 libcrypto.1.1.dylib`bread_conv + 32
    frame #3: 0x000000010bfc6299 libcrypto.1.1.dylib`bio_read_intern + 217
    frame #4: 0x000000010bfc6186 libcrypto.1.1.dylib`BIO_read + 38
    frame #5: 0x000000010bf111b2 libssl.1.1.dylib`ssl3_read_n + 610
    frame #6: 0x000000010bf142d9 libssl.1.1.dylib`ssl3_get_record + 185
    frame #7: 0x000000010bf13038 libssl.1.1.dylib`ssl3_read_bytes + 488
    frame #8: 0x000000010bf1cd1c libssl.1.1.dylib`ssl3_read_internal + 188
    frame #9: 0x000000010bf2a396 libssl.1.1.dylib`SSL_read + 38
    frame #10: 0x000000010c389c57 _ssl.cpython-37m-darwin.so`_ssl__SSLSocket_read + 903
    frame #11: 0x000000010aac6b02 libpython3.7m.dylib`_PyMethodDef_RawFastCallKeywords + 226
    frame #12: 0x000000010aacd161 libpython3.7m.dylib`_PyMethodDescr_FastCallKeywords + 81
    frame #13: 0x000000010aba2455 libpython3.7m.dylib`call_function + 869
    frame #14: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #15: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #16: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #17: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #18: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #19: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #20: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #21: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #22: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #23: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #24: 0x000000010aac7016 libpython3.7m.dylib`_PyObject_Call_Prepend + 150
    frame #25: 0x000000010aac5738 libpython3.7m.dylib`_PyObject_FastCallDict + 360
    frame #26: 0x000000010aac7eab libpython3.7m.dylib`object_vacall + 619
    frame #27: 0x000000010aac7bbd libpython3.7m.dylib`PyObject_CallMethodObjArgs + 173
    frame #28: 0x000000010ac42a37 libpython3.7m.dylib`_bufferedreader_raw_read + 151
    frame #29: 0x000000010ac42294 libpython3.7m.dylib`_buffered_readline + 756
    frame #30: 0x000000010ac441af libpython3.7m.dylib`_io__Buffered_readline + 95
    frame #31: 0x000000010aac6cb1 libpython3.7m.dylib`_PyMethodDef_RawFastCallKeywords + 657
    frame #32: 0x000000010aacd161 libpython3.7m.dylib`_PyMethodDescr_FastCallKeywords + 81
    frame #33: 0x000000010aba2455 libpython3.7m.dylib`call_function + 869
    frame #34: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #35: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #36: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #37: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #38: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #39: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #40: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #41: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #42: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #43: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #44: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #45: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #46: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #47: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #48: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #49: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #50: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #51: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #52: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #53: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #54: 0x000000010ab9f0bb libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29531
    frame #55: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #56: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #57: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #58: 0x000000010ab9effe libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29342
    frame #59: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #60: 0x000000010aac59d5 libpython3.7m.dylib`_PyFunction_FastCallDict + 501
    frame #61: 0x000000010ab9f26f libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29967
    frame #62: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #63: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #64: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #65: 0x000000010ab9f0bb libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29531
    frame #66: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #67: 0x000000010aac59d5 libpython3.7m.dylib`_PyFunction_FastCallDict + 501
    frame #68: 0x000000010ab9f26f libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29967
    frame #69: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #70: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #71: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #72: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #73: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #74: 0x000000010ab9f26f libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29967
    frame #75: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #76: 0x000000010aac5f06 libpython3.7m.dylib`_PyFunction_FastCallKeywords + 230
    frame #77: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #78: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #79: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #80: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #81: 0x000000010ab9ef53 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29171
    frame #82: 0x000000010aac643d libpython3.7m.dylib`function_code_fastcall + 253
    frame #83: 0x000000010aba2419 libpython3.7m.dylib`call_function + 809
    frame #84: 0x000000010ab9ef33 libpython3.7m.dylib`_PyEval_EvalFrameDefault + 29139
    frame #85: 0x000000010aba3175 libpython3.7m.dylib`_PyEval_EvalCodeWithName + 2981
    frame #86: 0x000000010ab97cc0 libpython3.7m.dylib`PyEval_EvalCode + 48
    frame #87: 0x000000010abdc8be libpython3.7m.dylib`PyRun_FileExFlags + 174
    frame #88: 0x000000010abdbdb5 libpython3.7m.dylib`PyRun_SimpleFileExFlags + 277
    frame #89: 0x000000010abff2df libpython3.7m.dylib`pymain_main + 6959
    frame #90: 0x000000010abffebd libpython3.7m.dylib`_Py_UnixMain + 125
    frame #91: 0x000000010a835f24 python3.7`start + 52
```
</details>

If I understand correctly, when talking to the server, lieer allows infinite timeout, which is never a good idea when doing network IO.

I guess I’ll try to capture stdout and stderr to see what is going on, and probably try to add a timeout next.

**Comment by gauteh** [comment]
I recently had the same problem and set the timeout to 180 seconds (gmi
set). I think the default should be something high, like 10 minutes:
#151, rather than infinite.

**Comment by gauteh** [comment]
Fixed in #151, you probably have to set this manually for existing repositories.
